File: Legacy/Leetcode/60PermutationSequence.py
# By listing and labeling all of the permutations in order,
# We get the following sequence (ie, for n = 3):
#
# "123"
# "132"
# "213"
# "231"
# "312"
# "321"
# Given n and k, return the kth permutation sequence.
#
# Note: Given n will be between 1 and 9 inclusive.

def get_permutation(n, k):
    not_taken = [True for _ in range(n)]
    index = k - 1
    table = [1]
    for i in range(1, n):
        table.append(table[-1] * i)
    ans = ''
    for i in range(n, 0, -1):
        pos_in_rest = index / table[i - 1] + 1
        index = index % table[i - 1]
        total = 0
        for i in range(n):
            total += not_taken[i]
            if total == pos_in_rest:
                ans += str(i + 1)
                not_taken[i] = False
                break
    return ans

print get_permutation(9, 5)

File: C++/Math/60. Permutation Sequence.cpp
class Solution {
public:
    
    string getPermutation(int n, int k) {
        string a = "";
        for (int i = 1; i <= n; i++) // 以n=9为例a初始化为“123456789”
            a += to_string(i);
        vector<int> fact(n + 1);
        fact[0] = 1;
        for (int i = 1; i <= n; i++) // 计算1到n的阶乘
            fact[i] = fact[i - 1] * i;
        k--;
        string ans = "";
        for (int i = n - 1; i >= 0; i--) { // 当前数字后面的数字个数：从n-1到0
            int index = k / fact[i]; // 后面有i个数字，组合为i!，当前k对i!的倍数决定当前数字
            k %= fact[i]; // 更新k为k对i!的余数
            ans += a[index];
            a.erase(a.begin() + index); // a删除对应数字
        }
        return ans;
    }
    
};